cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java

```diff
@@ -16,51 +16,46 @@
 
-import com.google.inject.Injector;
 import io.cdap.cdap.app.runtime.ProgramRuntimeProvider;
 import io.cdap.cdap.common.conf.CConfiguration;
-import io.cdap.cdap.internal.app.worker.ConfiguratorTask;
-import io.cdap.cdap.internal.app.worker.SystemAppTask;
+import io.cdap.cdap.common.lang.ClassLoaders;
 import io.cdap.cdap.master.environment.MasterEnvironments;
 import io.cdap.cdap.master.spi.environment.MasterEnvironment;
 import io.cdap.cdap.proto.ProgramType;
 import org.junit.Test;
+import org.junit.runner.RunWith;
 
 import java.util.ServiceLoader;
 
 /**
- * Test for {@link SparkProgramRuntimeProvider}.
+ * Test for {@link SparkProgramRuntimeProvider}. We need to use the delegated test runner to ensure that there aren't
+ * any casting errors for CDAP classes due to using different effective classloaders at runtime.
  */
+@RunWith(NoSparkClassLoaderTestRunner.class)
 public class SparkProgramRuntimeProviderTest {
 
   @Test
-  public void testConfiguratorTaskInjector() throws Exception {
+  public void testSparkResourceFilterStreamingClassLoading() throws Exception {
+    // Tests the scenario in which the parent classloader of the SparkProgramRuntimeProvider does not have access to
+    // Spark classes. We simulate this by wrapping the entire test class with a delegated classloader which filters out
+    // Spark classes.
+
+    // Set the thread context classloader as a delegate classloader which does not have the Spark classes.
+    // This simulates that the parent classloader of SparkProgramRuntimeProvider does not have Spark classes.
+    ClassLoader noSparkClassLoader = getClass().getClassLoader();
+    ClassLoader tmp = ClassLoaders.setContextClassLoader(noSparkClassLoader);
     MasterEnvironment masterEnvironment = new MockMasterEnvironment();
     masterEnvironment.initialize(null);
     MasterEnvironment tmpMasterEnv = MasterEnvironments.setMasterEnvironment(masterEnvironment);
 
     CConfiguration cConf = CConfiguration.create();
-    Injector injector = ConfiguratorTask.createInjector(cConf);
-    ServiceLoader<ProgramRuntimeProvider> serviceLoader = ServiceLoader.load(ProgramRuntimeProvider.class);
+    ServiceLoader<ProgramRuntimeProvider> serviceLoader = ServiceLoader.load(ProgramRuntimeProvider.class,
+                                                                             noSparkClassLoader);
     SparkProgramRuntimeProvider runtimeProvider = (SparkProgramRuntimeProvider) serviceLoader.iterator().next();
-    runtimeProvider.createProgramClassLoader(cConf, ProgramType.SPARK);
-
+    ClassLoader sparkClassLoader = runtimeProvider.createProgramClassLoader(cConf, ProgramType.SPARK);
+    // Load any spark streaming class.
+    sparkClassLoader.loadClass("org.apache.spark.streaming.StreamingContext");
     MasterEnvironments.setMasterEnvironment(tmpMasterEnv);
+    ClassLoaders.setContextClassLoader(tmp);
   }
-
-  @Test
-  public void testSystemAppTaskInjector() throws Exception {
```

Comment:
Why removing this test?

Reply:
When ArtifactClassLoaderFactory originally used createProgramRunner, it needed to be able to dynamically instantiate a [Distributed]SparkProgramRunner using the existing Guice bindings; thus, this test was added to ensure that the Guice bindings are always present.

However, after refactoring the ArtifactClassLoaderFactory to remove this logic, I don't think we need these tests anymore since we no longer instantiate the entire ProgramRunner, and the injector is not used in the `createProgramClassLoader` contract.

Reply:
The problem is that the `SystemAppTask.createInjector` method is still there and in used. This method is to test that, regardless what changed inside the guice bindings for that injector.

Reply:
`SystemAppTask.createInjector` is separately tested in [SystemAppTaskTest.java](https://github.com/cdapio/cdap/blob/2d2ae49d40d87b96a72cbeaa767561a8ae1a041c/cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/worker/SystemAppTaskTest.java). This test is specifically intended for testing that `ProgramRunner` can be properly instantiated by a SystemAppTask injector; however, the `SystemAppTask.createInjector` method is no longer used to instantiate a program runner anywhere.

Reply:
I see. Ok to remove. BTW, we should all the tests against the same class in the same test class if possible.

Reply:
Yup; main reason I had to have it in this class in the first place is because I'm not sure how to get the runtime extension classes (i.e. SparkProgramRuntimeProvider) in appfabric. However, will keep this in mind in the future. Thanks!